[Serialization] Improve module loading performance

ahoppen opened the pull request:
When looking for a Swift module on disk, we were scanning all module search paths if they contain the module we are searching for. In a setup where each module is contained in its own framework search path, this scaled quadratically with the number of modules being imported. E.g. a setup with 100 modules being imported form 100 module search paths could cause on the order of 10,000 checks of `FileSystem::exists`. While these checks are fairly fast (~10µs), they add up to ~100ms.

To improve this, perform a first scan of all module search paths and list the files they contain. From this, create a lookup map that maps filenames to the search paths they can be found in. E.g. for
```
searchPath1/
  Module1.framework

searchPath2/
  Module1.framework
  Module2.swiftmodule
```
we create the following lookup table
```
Module1.framework -> [searchPath1, searchPath2]
Module2.framework -> [searchPath2]
```

The initial scan is fairly fast (in O(files inside module search paths)) and afterwards finding a module in the framework search paths is in O(1) in the situation described above.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

xymus commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -599,8 +683,13 @@ SerializedModuleLoaderBase::findModule(ImportPath::Element moduleID,
+  SmallVector<std::string, 3> InterestingFilenames = {
+      (moduleName + ".framework").str(),
+      genericBaseName.getName(file_types::TY_SwiftModuleInterfaceFile),
+      genericBaseName.getName(file_types::TY_SwiftModuleFile)};
```
Should we also check for `TY_PrivateSwiftModuleInterfaceFile` here?

ahoppen replied:
I checked that it wasn’t necessary because both places that can consume the private swift module interface bail out if the public one doesn’t exist.

https://github.com/apple/swift/blob/2ef512722ae8bbcc7dff272a9db99e1dd8a33752/lib/Frontend/ModuleInterfaceLoader.cpp#L1108
bails out if the public file doesn’t exist here
https://github.com/apple/swift/blob/2ef512722ae8bbcc7dff272a9db99e1dd8a33752/lib/Frontend/ModuleInterfaceLoader.cpp#L1097

---

https://github.com/apple/swift/blob/2ef512722ae8bbcc7dff272a9db99e1dd8a33752/lib/Serialization/ModuleDependencyScanner.cpp#L61
asserts that the public file exists here
https://github.com/apple/swift/blob/2ef512722ae8bbcc7dff272a9db99e1dd8a33752/lib/Serialization/ModuleDependencyScanner.cpp#L56

xymus replied:
That's interesting I didn't realize we had this restriction in place. Thank you for looking into it.

ahoppen replied:
Just to be safe (also for the future if we decide to lift that restriction), I added the private swiftinterface file as well.

xymus reviewed (approved):
This sounds like an overdue improvement, thanks!

bnbarham commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -82,12 +82,6 @@ void forEachTargetModuleBasename(const ASTContext &Ctx,
-
 /// Apply \p body for each module search path in \p Ctx until \p body returns
 /// non-None value. Returns the return value from \p body, or \c None.
 Optional<bool> forEachModuleSearchPath(
```
The only other use of `forEachModuleSearchPath` is in `collectVisibleTopLevelModuleNamesImpl` and at a cursory glance it seems like we could use `forEachModuleSearchPathContainingFile` there as well.

ahoppen replied:
I don’t think we can because we don’t know what we’re looking for. 
https://github.com/apple/swift/blob/2ef512722ae8bbcc7dff272a9db99e1dd8a33752/lib/Serialization/SerializedModuleLoader.cpp#L189-L206
just enumerates over all files in the search path.

We could probably refactor that method to recover the first level directory structure from `RelevantSearchPathForFilename` but I don’t think it’s worth it.

bnbarham replied:
Not sure how I missed the `forEachDirectyEntryPath`, that's what I get for skimming over it quickly. Bit unfortunate but it's just doing it once, so I guess it's fine 👍

bnbarham reviewed (commented):
Nice! Those microseconds really add up 😆

ahoppen commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -166,9 +168,10 @@ Status ModuleFile::associateWithFileContext(FileUnit *file, SourceLoc diagLoc,
-    ctx.addSearchPath(
-        ctx.SearchPathOpts.SearchPathRemapper.remapPath(searchPath.Path),
-        searchPath.IsFramework, searchPath.IsSystem);
+    auto Path =
```
I hate our inconsistency with capitalization of variables 😬.

bnbarham replied:
... yeah. Honestly I never know whether to comment about it or not.

ahoppen commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -123,6 +117,95 @@ Optional<bool> forEachModuleSearchPath(
+void SerializedModuleLoaderBase::scanFilesInSearchPath(
+    llvm::vfs::FileSystem &FS, StringRef SearchPath, SearchPathKind Kind,
+    bool IsSystem) {
+  if (!HasPopulatedRelevantSearchPathsForFile) {
+    // If we have not populated RelevantSearchPathsForFile yet, the new search
+    // path will be picked up by populateRelevantSearchPathsForFileIfNecessary
+    // once its called. So nothing to do here.
+    return;
+  }
```
If we didn’t have this check we could end up in the following situation: We have import paths `A` and `B`. Before `RelevantSearchPathsForFile` is populated we discover import path `C` through a module import which would thus be added to `RelevantSearchPathsForFile`. When we call `populateRelevantSearchPathsForFileIfNecessary` we load `A, B, C` from the `ASTContext` and end up with with import search paths `C, A, B, C` (in that order) in `RelevantSearchPathsForFile`, but `A, B, C` is correct.

Also added the comment to the source code for future reference.

bnbarham replied:
Ah, I didn't realise that the search paths were added to the `ASTContext` as well. Fair enough, hopefully we rip out adding search paths from modules as we go at some point anyway!

ahoppen commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -123,6 +117,95 @@ Optional<bool> forEachModuleSearchPath(
+  // search paths have already been added to RelevantSearchPaths by their index
+  // in AddedRelevantSearchPathsIndicies.
+  SmallVector<const SearchPath *, 4> RelevantSearchPaths;
+  std::set<unsigned> RelevantSearchPathsIndicies;
```
Good idea. I didn’t know about `SmallSet`.

ahoppen commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -123,6 +117,95 @@ Optional<bool> forEachModuleSearchPath(
+  // different filenames.
+  llvm::sort(RelevantSearchPaths);
+
+  for (auto &Entry : RelevantSearchPaths) {
+    if (auto result = callback(Entry->Path, Entry->Kind, Entry->IsSystem)) {
+      return result;
+    }
+  }
```
That’s a very good idea. 👍

ahoppen commented:
OK, I needed to do a pretty big refactoring of the PR because before each module loader maintained its own lookup table. Thus, if one module loader caused a search path to get added, the others wouldn’t know about it. So I moved the entire lookup logic down to `SearchPathOpts`, which has a better overview of any search path modifications anyway.

ahoppen commented:
https://github.com/apple/llvm-project/pull/3546

@swift-ci Please smoke test

bnbarham commented on lib/AST/SearchPathOptions.cpp:
```diff
@@ -0,0 +1,96 @@
+       !Error && Dir != llvm::vfs::directory_iterator(); Dir.increment(Error)) {
+    StringRef Filename = llvm::sys::path::filename(Dir->path());
+    LookupTable[Filename].push_back(
+        {SearchPath.str(), Kind, IsSystem, SearchPathIndex});
```
If we store the results of `getDarwinImplicitFrameworkSearchPaths ` in `SearchPathOptions` this could then be a `StringRef` (I think). Possibly worth it?

ahoppen replied:
With the current design, we only decide whether we need these paths (i.e. whether we are on a Darwin platform) when enumerating the search paths. So if we want to store the Darwin implicit framework search paths in `SearchPathOptions`, we would need to always compute them when setting SDK path, regardless of the platform (or have a slightly more complicated caching method). I’m not sure if it’s worth it. My current gut feeling is that this function will be called <10 times per compilation/SourceKit operation.

The clean solution would be that `SearchPathOptions` knows whether it’s on a Darwin platform, but I think that’s out of scope for this PR.

bnbarham replied:
Just seemed like a relatively large map to me, but I guess there's not *that* many file -> search paths. I was mostly hoping we could move the platform check into `ParseSearchPathArgs` and just add the extra paths directly to the framework search paths rather than have separate field, but I haven't looked into that at all. If that's not possible then I don't really have a preference either way 🤷

ahoppen replied:
Ah, sorry. I misunderstood you. I though you wanted to avoid the computation of `getDarwinImplicitFrameworkSearchPaths`. Making `ModuleSearchPath` only contain a `StringRef` does seem like a worthwhile improvement 👍 

> I was mostly hoping we could move the platform check into ParseSearchPathArgs and just add the extra paths directly to the framework search paths rather than have separate field, but I haven't looked into that at all. If that's not possible then I don't really have a preference either way 🤷

We can’t do that because framework search paths may be added after the compiler invocation is created and those should come before the Darwin implicit search paths.

bnbarham replied:
Ah I see. Yeah the computation didn't really bother me at all - from a quick look this just seemed to be the only part that was preventing using StringRefs.

bnbarham commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -177,12 +177,12 @@ namespace {
       }
 
-      for (auto path : searchPathOpts.ImportSearchPaths) {
+      for (auto path : searchPathOpts.getImportSearchPaths()) {
```
I know you only changed to the getter here, but it's always nice to be explicit with `const auto &`. There's a few others where the same applies.

bnbarham commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -593,74 +589,98 @@ SerializedModuleLoaderBase::findModule(ImportPath::Element moduleID,
+      isFramework = false;
+
+      // On Apple platforms, we can assume that the runtime libraries use
+      // target-specifi module files wihtin a `.swiftmodule` directory.
```
I know you didn't actually change any of this but:
target-specifi -> target-specific
wihtin -> within

bnbarham commented on lib/AST/SearchPathOptions.cpp:
```diff
@@ -0,0 +1,96 @@
+    rebuildLookupTable(Opts, FS, IsOSDarwin);
+  }
+
+  // Gather all search paths that include a file whose name is in Filenames.
+  // To make sure that we don't include search paths twice, keep track of which
+  // search paths have already been added to Result by their kind and Index
+  // in ResultIds.
+  llvm::SmallVector<const ModuleSearchPath *, 4> Result;
+  llvm::SmallSet<std::pair<ModuleSearchPathKind, unsigned>, 4> ResultIds;
```
We can still add a search path twice now right?

ahoppen replied:
Yes, if it’s specified twice on the command line (and not cleaned up earlier). Is that good or bad or are you just wondering? Search paths could also exist twice as different kinds, e.g. once as framework search path and once as import search path.

bnbarham replied:
I was mostly just pointing out `To make sure that we don't include search paths twice` but it is clarified later in that sentence so 🤷

ahoppen replied:
Oh, yes. Good catch.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/llvm-project/pull/3546

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test macOS

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please test source compatibility

ahoppen committed:
[Serialization] Improve module loading performance

When looking for a Swift module on disk, we were scanning all module search paths if they contain the module we are searching for. In a setup where each module is contained in its own framework search path, this scaled quadratically with the number of modules being imported. E.g. a setup with 100 modules being imported form 100 module search paths could cause on the order of 10,000 checks of `FileSystem::exists`. While these checks are fairly fast (~10µs), they add up to ~100ms.

To improve this, perform a first scan of all module search paths and list the files they contain. From this, create a lookup map that maps filenames to the search paths they can be found in. E.g. for
```
searchPath1/
  Module1.framework

searchPath2/
  Module1.framework
  Module2.swiftmodule
```
we create the following lookup table
```
Module1.framework -> [searchPath1, searchPath2]
Module2.swiftmodule -> [searchPath2]
```

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3546

@swift-ci Please smoke test

nkcsgexi commented:
@ahoppen sorry that I have to revert it despite this is a very nice change: https://github.com/apple/swift/pull/40663
